File: src/main/java/me/testaccount666/serversystem/placeholderapi/executables/BalancePlaceholder.java
package me.testaccount666.serversystem.placeholderapi.executables;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.placeholderapi.Placeholder;
import me.testaccount666.serversystem.userdata.OfflineUser;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;
import java.util.Set;
import java.util.UUID;

public class BalancePlaceholder implements Placeholder {
    @Override
    public String execute(@Nullable OfflineUser user, String identifier, String... arguments) {
        if (user == null && arguments.length == 0) return "No user specified!";
        if (user == null || arguments.length >= 1) {
            var player = getOfflinePlayer(arguments[0]);
            var userOptional = getOfflineUser(player);
            if (userOptional.isEmpty()) return "User ${arguments[0]}} not found!";

            user = userOptional.get();
        }

        var bankAccount = user.getBankAccount();
        var balance = bankAccount.getBalance();

        var formatBalance = identifier.equalsIgnoreCase("balance");
        if (!formatBalance) return String.format("%.2f", balance.doubleValue());

        return ServerSystem.Instance.getEconomyProvider().formatMoney(balance);
    }

    @Override
    public Set<String> getIdentifiers() {
        return Set.of("balance", "unformattedbalance");
    }

    private OfflinePlayer getOfflinePlayer(String nameOrUuid) {
        try {
            var uuid = UUID.fromString(nameOrUuid);
            return Bukkit.getOfflinePlayer(uuid);
        } catch (IllegalArgumentException ignored) {
        }
        return Bukkit.getOfflinePlayer(nameOrUuid);
    }

    private Optional<OfflineUser> getOfflineUser(OfflinePlayer player) {
        var userOptional = ServerSystem.Instance.getUserManager().getUser(player.getUniqueId());
        if (userOptional.isEmpty()) return Optional.empty();
        var cachedUser = userOptional.get();
        return Optional.of(cachedUser.getOfflineUser());
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/placeholderapi/executables/BalancePlaceholder.kt
package me.testaccount666.serversystem.placeholderapi.executables

import me.testaccount666.serversystem.extensions.format
import me.testaccount666.serversystem.extensions.getService
import me.testaccount666.serversystem.placeholderapi.Placeholder
import me.testaccount666.serversystem.userdata.OfflineUser
import me.testaccount666.serversystem.userdata.UserManager
import me.testaccount666.serversystem.userdata.money.EconomyProvider
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import java.util.*

class BalancePlaceholder : Placeholder {
    override fun execute(user: OfflineUser?, identifier: String, vararg arguments: String): String {
        var user = user
        if (user == null && arguments.isEmpty()) return "No user specified!"
        if (user == null || arguments.isNotEmpty()) {
            val player = getOfflinePlayer(arguments[0])
            val targetUser = getOfflineUser(player) ?: return "User ${arguments[0]} not found!"

            user = targetUser
        }

        val balance = user.bankAccount.balance

        val formatBalance = identifier.equals("balance", true)
        if (!formatBalance) return balance.toDouble().format("%.2f")

        return getService<EconomyProvider>().formatMoney(balance)
    }

    override val identifiers = setOf("balance", "unformattedbalance")

    private fun getOfflinePlayer(nameOrUuid: String): OfflinePlayer {
        try {
            return Bukkit.getOfflinePlayer(UUID.fromString(nameOrUuid))
        } catch (_: IllegalArgumentException) {
        }
        return Bukkit.getOfflinePlayer(nameOrUuid)
    }

    private fun getOfflineUser(player: OfflinePlayer): OfflineUser? {
        val cachedUser = getService<UserManager>().getUserOrNull(player.uniqueId)
        return cachedUser?.offlineUser
    }
}
